### PR TITLE
Rd-gen

### PR DESCRIPTION
**Pull request by Domonion**
# Description

Moved child process communication to RD. Added ability to track process through JetBrains.Lifetimes. Refactored ConcreteExecutor to support JetBrains.Lifetimes and RD

1. switched from Protocol.Command to rd generated model
2. added utbot-rd project featuring LifetimedProcess, ProcessWithRdServer and minor rd-related utility classes
3. Switched from message-driven ConcreteExecutor to callback rd model, featuring UtInstrumentationProcess

## Type of Change

Please delete options that are not relevant.

- Refactoring (typos and non-functional changes) 
- New feature (non-breaking change which adds functionality)

## Automated Testing

org.utbot.examples.collections.ListsTest
org.utbot.examples.collections.MapsTest
org.utbot.external.api.UtBotJavaApi
org.utbot.examples.exceptions.JvmCrashExamplesTest
All tests should work

### CODE REVIEW

File: utbot-analytics/build.gradle [hunk 1]
```diff
@@ -82,4 +82,5 @@ jar {
     }
 
     duplicatesStrategy = DuplicatesStrategy.EXCLUDE
+    zip64 = true
```
Comment (korifey) [inline utbot-analytics/build.gradle:85]: why?

File: utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/process/ChildProcessRunner.kt [hunk 2]
```diff
@@ -38,20 +38,19 @@ class ChildProcessRunner {
...
-            errorLogFile = File(UT_BOT_TEMP_DIR, "${hashCode()}-${processSeqN}.log")
+            errorLogFile = File(UT_BOT_TEMP_DIR, "${id}-${processSeqN}.log")
         }
 
         val directory = WorkingDirService.provide().toFile()
-
-        val processBuilder = ProcessBuilder(cmds)
-            .redirectError(errorLogFile)
-            .directory(directory)
+        val processBuilder = ProcessBuilder(cmds + " " + portArgument).redirectError(errorLogFile).directory(directory)
```
Comment (SBOne-Kenobi) [inline utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/process/ChildProcessRunner.kt:53]: Empty argument (`" "`) is needed?

File: utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/process/ChildProcess.kt [hunk 3]
```diff
@@ -39,152 +58,202 @@ internal object HandlerClassesLoader : URLClassLoader(emptyArray()) {
...
-                } catch (e: Throwable) {
-                    log("Terminating process because exception occured: ${e.stackTraceToString()}")
-                    exitProcess(1)
+    GlobalScope.launchChild(Lifetime.Eternal, Dispatchers.Unconfined) {
+        while (true) {
+            val now = System.currentTimeMillis()
+            val start = executionStart.get()
+            val end = executionEnd.get()
+
+            if (start > end) { // process is doing something
```
Comment (SBOne-Kenobi) [inline utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/process/ChildProcess.kt:114]: Could you use conditions and mutexes for waiting?

File: utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/process/ChildProcess.kt [hunk 4]
```diff
@@ -39,152 +58,202 @@ internal object HandlerClassesLoader : URLClassLoader(emptyArray()) {
...
+                delay(1000)
+            } else { // process is waiting for message
+                if (now - end > messageFromMainTimeoutMillis) {
+                    logInfo { "terminating lifetime" }
+                    def.terminate()
+                    break
+                } else {
+                    delay(1000)
```
Comment (SBOne-Kenobi) [inline utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/process/ChildProcess.kt:122]: Magic number

File: utbot-analytics/build.gradle (hunk 1 above)
Comment (Domonion) [reply utbot-analytics/build.gradle:85]: > Execution failed for task ':utbot-analytics:jar'.
> > archive contains more than 65535 entries.
Without this flag gradle cannot build it

File: utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/process/ChildProcess.kt [hunk 5]
```diff
@@ -40,158 +59,207 @@ internal object HandlerClassesLoader : URLClassLoader(emptyArray()) {
...
+    logInfo { "kryo created" }
+
+    val clientProtocol = UtRdUtil.createUtClientProtocol(lifetime, port, UtSingleThreadScheduler { logInfo(it) })
+    logInfo {
+        "heartbeatAlive - ${clientProtocol.wire.heartbeatAlive.value}, connected - ${
+            clientProtocol.wire.connected.value
+        }"
+    }
+    val (sync, protocolModel) = obtainClientIO(lifetime, clientProtocol)
+    protocolModel.warmup.set { _ ->
```
Comment (SBOne-Kenobi) [inline utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/process/ChildProcess.kt:191]: Can you carry out method with setting up commands for protocol model?

File: utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/ConcreteExecutor.kt [hunk 6]
```diff
@@ -132,13 +134,85 @@ class ConcreteExecutor<TIResult, TInstrumentation : Instrumentation<TIResult>> p
...
+                    pathsToDependencyClasses,
+                    classLoader
+                )
+                processInstance = proc
+            }
+        }
+
+        return proc!!
+    }
+    suspend fun <T> withProcess(exclusively: Boolean = false, block: suspend UtInstrumentationProcess.() -> T): T {
```
Comment (SBOne-Kenobi) [inline utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/ConcreteExecutor.kt:168]: Provide description

**Review by SBOne-Kenobi — CHANGES_REQUESTED** [review]
Formatting questions

I hope that you will write one or several docs or at least a brief description in the PR about:
- How communication between child and main processes works
- How to support new commands, some requirements and tips for this
- A brief description about ChildProcess structure and workflow
- A brief description about rdgen builds

File: utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/process/ChildProcess.kt [hunk 7]
```diff
@@ -84,10 +85,12 @@ private fun logTrace(any: () -> Any?) {
...
 
-private val executionStart = AtomicLong(1)
-private val executionEnd = AtomicLong(0)
-private const val messageFromMainTimeoutMillis = 120 * 1000
-private const val delayTimeoutMillis = 1000L
+private enum class State {
+    STARTED,
+    ENDED
+}
+private const val messageFromMainTimeoutMillis: Long = 120 * 1000
```
Comment (SBOne-Kenobi) [inline utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/process/ChildProcess.kt:92]: Magic constant `1000`, TimeUnit.Seconds.toMillis may help

**Review by SBOne-Kenobi — APPROVED** [review]

**Pull request merged**